cql3/statements/create_keyspace_statement.cc

```diff
@@ -134,50 +134,58 @@ future<> cql3::statements::create_keyspace_statement::grant_permissions_to_creat
 // if the strategy is forbidden by the configuration, or return a warning
-// string if the restriction was set to "warn".
+// string if the restriction was set to warning level.
 // This function is only supposed to check for replication strategies
 // restricted by the configuration. Checks for other types of strategy
 // errors (such as unknown replication strategy name or unknown options
 // to a known replication strategy) are done elsewhere.
-std::optional<sstring> check_restricted_replication_strategy(
+std::vector<sstring> check_against_restricted_replication_strategies(
     query_processor& qp,
     const sstring& keyspace,
     const ks_prop_defs& attrs,
     cql_stats& stats)
 {
     if (!attrs.get_replication_strategy_class()) {
-        return std::nullopt;
+        return {};
     }
-    sstring replication_strategy = locator::abstract_replication_strategy::to_qualified_class_name(
-        *attrs.get_replication_strategy_class());
-    auto& topology = qp.proxy().get_token_metadata_ptr()->get_topology();
-    // SimpleStrategy is not recommended in any setup which already has - or
-    // may have in the future - multiple racks or DCs. So depending on how
-    // protective we are configured, let's prevent it or allow with a warning:
-    if (replication_strategy == "org.apache.cassandra.locator.SimpleStrategy") {
-        switch(qp.db().get_config().restrict_replication_simplestrategy()) {
-        case db::tri_mode_restriction_t::mode::TRUE:
-            throw exceptions::configuration_exception(
-                "SimpleStrategy replication class is not recommended, and "
-                "forbidden by the current configuration. Please use "
-                "NetworkToplogyStrategy instead. You may also override this "
-                "restriction with the restrict_replication_simplestrategy=false "
-                "configuration option.");
-        case db::tri_mode_restriction_t::mode::WARN:
-            return format("SimpleStrategy replication class is not "
-                "recommended, but was used for keyspace {}. The "
-                "restrict_replication_simplestrategy configuration option "
-                "can be changed to silence this warning or make it into an error.",
-                keyspace);
-        case db::tri_mode_restriction_t::mode::FALSE:
+
+    std::vector<sstring> warnings;
+    auto replication_strategy = locator::abstract_replication_strategy::create_replication_strategy(
+            locator::abstract_replication_strategy::to_qualified_class_name(
+                    *attrs.get_replication_strategy_class()), {})->get_type();
+    auto rs_warn_list = qp.db().get_config().replication_strategy_warn_list();
+    auto rs_fail_list = qp.db().get_config().replication_strategy_fail_list();
+
+    if (replication_strategy == locator::replication_strategy_type::simple) {
+        if (auto simple_strategy_restriction = qp.db().get_config().restrict_replication_simplestrategy();
```

Comment:
I think that the (now legacy) restrict_replication_simplestrategy option must be handled in main() by adding the relevant entry into the rs_fail_list or rs_warn_list. And _this_ place just checks both lists

Reply:
From my point of view, this was the only place where `restrict_replication_simplestrategy` was being used, hence I brought it back and am reimplementing how it used to work with the new "mechanism". I don't think there would be any observable differences between inserting `restrict_replication_simplestrategy` to the warn/fail lists in main instead of `create_keyspace_statement.cc`. other than reading it once (main case) vs reading it every time (current case). But reading it on every event isn't necessarily a bad thing, I imagine, because someone could have changed `restrict_replication_simplestrategy` via system.config virtual table/API/sending signal to reload scylla.yaml and expects `restrict_replication_simplestrategy` behaviour to change. That's actually how it was working, to the best of my knowledge.
But maybe I am missing something, so could you please elaborate on your motivation?

Reply:
> From my point of view, this was the only place where restrict_replication_simplestrategy was being used, hence I brought it back and am reimplementing how it used to work with the new "mechanism"

Yes, this makes perfect sense.

> But maybe I am missing something, so could you please elaborate on your motivation?

Currently there's a ... strange bug in your approach. Consider someone (maybe mistakenly) uses both option -- `--restrict_replication_simplestrategy=warn --replication-strategy-warn-list=simple`. In that case simple strategy would get into function-local warn-list twice (first naturally, second due to this fallback). Respectively, there will be _two_ warnings on the screen. Weird result of weird usage? Yes, definitely. Were that fallback performed once on boot, we could've checked if the simple strategy is on the new list and not re-emplace-back it. Doing it here is a bit of an overkill.

Reply:
Two simplestrategies would be inserted into the warn list, yes, but that would end up generating only 1 warning message (there's no `for` loop anywhere in the code I introduce, just an `if` to see if a currently-used strategy is present in any list, and it doesn't matter if an entry in a list is duplicated), please consider:
```
cqlsh> select * from system.config where name = 'restrict_replication_simplestrategy';

 name                                | source | type             | value
-------------------------------------+--------+------------------+--------
 restrict_replication_simplestrategy |    cql | restriction mode | "warn"

(1 rows)
cqlsh> select * from system.config where name = 'replication_strategy_warn_list';

 name                           | source  | type                      | value
--------------------------------+---------+---------------------------+--------------------
 replication_strategy_warn_list | default | replication strategy list | ["SimpleStrategy"]

(1 rows)
cqlsh> CREATE KEYSPACE tracking42 WITH REPLICATION = { 'class' : 'SimpleStrategy', 'replication_factor': 3};

Warnings :
SimpleStrategy replication class is not recommended, but was used for keyspace tracking42. You may suppress this warning by delisting SimpleStrategy from replication_strategy_warn_list configuration option, or make it into an error by listing this replication strategy on replication_strategy_fail_list.

cqlsh> 
```
But what you wrote makes me think that maybe we should generate 2 warnings, one for each flag? OTOH not necessarily, because it's us who decide how the newly introduced guardrail behaves and we could as easily decide to not print a 2nd warning if the legacy option is on.

Reply:
> Two simplestrategies would be inserted into the warn list, yes, but that would end up generating only 1 warning message (there's no for loop anywhere in the code I introduce, just an if to see if a currently-used strategy is present in any list, and it doesn't matter if an entry in a list is duplicated)

Indeed. Then I withdraw this concern :)